python/ray/dag/gradio_visualization.py

```diff
@@ -0,0 +1,160 @@
+            node: the DAGNode to process
+            depths: map between DAGNode uuid to the current longest found distance
+                between the node and any InputAttributeNode
+
+        Returns:
+            The original node.
+        """
+        uuid = node.get_stable_uuid()
+        for child_node in node._get_all_child_nodes():
+            depths[uuid] = max(depths[uuid], depths[child_node.get_stable_uuid()] + 1)
+
+        self.uuid_to_node[uuid] = node
+        return node
+
+    async def _get_result(self, node_uuid: str):
+        """Retrieves the execution output of the inputted DAGNode, from last execution.
+
+        This function should only be called after a request has been sent through
+        self.handle.predict.remote() separately.
+        """
+
+        return await asyncio.wait_for(
+            await self.handle.get_object_ref_for_node.remote(node_uuid), timeout=20
+        )
+
+    def visualize_with_gradio(
+        self,
+        driver_handle: RayServeHandle,
+        port: Optional[int] = None,
+        _launch: bool = True,
+    ):
+        """
+        Launches a Gradio UI that allows interactive request dispatch and displays
+        the evaluated outputs of each node in a deployment graph in real time.
+
+        Args:
+            driver_handle: The handle to a DAGDriver deployment obtained through a call
+                to serve.run(). The DAG rooted at that DAGDriver deployment will be
+                visualized through Gradio.
+            _launch: Whether to launch the Gradio app. Used for unit testing purposes.
+        """
+
+        self._reset_state()
+        self.handle = driver_handle
+
+        # Load the root dag node from handle
+        dag_node_json = ray.get(self.handle.get_dag_node_json.remote())
+        dag = json.loads(dag_node_json, object_hook=dagnode_from_json)
+
+        # Get name and level for each node in dag
+        depths = defaultdict(lambda: 0)
+        dag.apply_recursive(lambda node: self._fetch_depths(node, depths))
+
+        # Wraps _get_result because functools.partial doesn't work with asynchronous
+        # class methods: https://stackoverflow.com/q/67020609/11162437
+        async def get_result_wrapper(node_uuid):
+            return await self._get_result(node_uuid)
+
+        with gr.Blocks() as demo:
+            self._make_blocks(depths)
+
+            submit = gr.Button("Submit")
+            # Add event listener that sends the request to the deployment graph
+            submit.click(
+                fn=lambda *args: self.handle.predict.remote(args),
+                inputs=list(self.input_index_to_block.values()),
+                outputs=[],
+            )
+            # Add event listeners that resolve object refs for each of the nodes
+            for node_uuid, block in self.uuid_to_block.items():
+                submit.click(partial(get_result_wrapper, node_uuid), [], block)
+
+        if _launch:
+            demo.launch(server_port=port)
```

Comment:
can we add a "clear" button to clear out the state? (good for demo to avoid intermediate output)